src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs

```diff
@@ -76,27 +83,46 @@ public MongoTypeMappingSource(TypeMappingSourceDependencies dependencies)
             var elementMapping = FindPrimitiveMapping(elementMappingInfo) ?? FindCollectionMapping(elementMappingInfo);
             return elementMapping == null
                 ? null
-                : new MongoTypeMapping(clrType, ComparerFactory.CreateComparer(elementMapping, clrType));
+                : new MongoTypeMapping(clrType, CreateComparer(elementMapping, clrType));
         }
 
         // Support collection on the entity
-        if (clrType.IsGenericType && !clrType.IsGenericTypeDefinition)
+        if (clrType is {IsGenericType: true, IsGenericTypeDefinition: false})
         {
             var genericTypeDefinition = clrType.GetGenericTypeDefinition();
             if (genericTypeDefinition == typeof(List<>)
                 || genericTypeDefinition == typeof(IList<>)
-                || genericTypeDefinition == typeof(ICollection<>)
                 || genericTypeDefinition == typeof(IReadOnlyList<>)
-                || genericTypeDefinition == typeof(IReadOnlyCollection<>))
+                || genericTypeDefinition == typeof(ObservableCollection<>)
+                || genericTypeDefinition == typeof(Collection<>))
             {
                 var elementMappingInfo = new TypeMappingInfo(itemType);
                 var elementMapping = FindPrimitiveMapping(elementMappingInfo) ?? FindCollectionMapping(elementMappingInfo);
                 return elementMapping == null
                     ? null
-                    : new MongoTypeMapping(clrType, ComparerFactory.CreateComparer(elementMapping, clrType));
+                    : new MongoTypeMapping(clrType, CreateComparer(elementMapping, clrType));
             }
         }
 
         return null;
     }
+
+    public static ValueComparer? CreateComparer(CoreTypeMapping elementMapping, Type collectionType)
+    {
+        // Only single dimension arrays are supported
+        if (collectionType.IsArray && collectionType.GetArrayRank() != 1)
+        {
+            return null;
+        }
+
+        var elementType = collectionType.TryGetItemType(typeof(IEnumerable<>))!;
+
+        return (ValueComparer?)Activator.CreateInstance(
+            elementType.IsNullableValueType()
+                ? typeof(NullableValueTypeListComparer<>).MakeGenericType(Nullable.GetUnderlyingType(elementType) ?? elementType)
```

Comment:
If it takes more then one ternary operator to implement a logic, I prefer to have an good-old `if` instead for better readability.

Reply:
With this approach it's clear that from line 120 onwards it's going to create an instance of something, and that it's dependent on whether it's a nullable value type and if not whether it is assignable or not.

With an if condition you'd have to trawl through 3-4 if statements to find out they all do the same thing with one of three different comparers. It's optimized for scanning through code quickly - the idea being that by the time you've read the first line you can decide whether this is the line you cared about or not - vs reading 15 lines of if conditions only to find they all return a valuecomparer but that wasn't what you were looking for/cared about.